app/assets/javascripts/3dbio_viewer/src/data/repositories/SourcesApiRepository.ts

```diff
@@ -0,0 +1,39 @@
+import { Codec, GetType, string } from "purify-ts";
+import { FutureData } from "../../domain/entities/FutureData";
+import { Source } from "../../domain/entities/Source";
+import { SourcesRepository } from "../../domain/repositories/SourcesRepository";
+import { routes } from "../../routes";
+import { getValidatedJSON } from "../request-utils";
+import { Pagination, getResults, paginationCodec } from "../codec-utils";
+
+export class SourcesApiRepository implements SourcesRepository {
+    get(): FutureData<Source[]> {
+        const { bionotesStaging: api } = routes;
+
+        const nmr$ = getValidatedJSON<Pagination<NmrMethod>>(
```

Comment:
Note that you don't need to pass the type, TS is very smart and takes it from the second argument (the codec).